File: bin/i18n/json-to-lua.lua
--[[
 * Copyright (C) Rotorflight Project
 *
 * License GPLv3: https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
]]

local json = dofile("lib/dkjson.lua")

local jsonRoot = "json"
local outRoot = "../../scripts/rfsuite/i18n"
local isWindows = package.config:sub(1,1) == "\\"

local function readHeader(path)
    local f = io.open(path, "r")
    if not f then return "" end
    local content = f:read("*a")
    f:close()
    return content .. "\n\n"
end

local fileHeader = readHeader("lib/header.txt")

-- Helper: list files/dirs
local function listDir(path)
    local cmd = isWindows
        and ('dir /b "%s" 2>nul'):format(path)
        or ('ls -1 "%s" 2>/dev/null'):format(path)
    local pipe = io.popen(cmd)
    local result = {}
    for line in pipe:lines() do table.insert(result, line) end
    pipe:close()
    return result
end

-- Helper: is directory?
local function isDir(path)
    local cmd = isWindows
        and ('if exist "%s\\" (echo d)'):format(path)
        or ('[ -d "%s" ] && echo d'):format(path)
    local pipe = io.popen(cmd)
    local result = pipe:read("*a")
    pipe:close()
    return result:match("d")
end

-- Ensure output directory exists
local function ensureDir(path)
    local cmd = isWindows
        and ('mkdir "%s" >nul 2>nul'):format(path)
        or ('mkdir -p "%s" 2>/dev/null'):format(path)
    os.execute(cmd)
end

-- Deep merge two nested tables
local function deepMerge(base, new)
    for k, v in pairs(new) do
        if type(v) == "table" and type(base[k]) == "table" then
            deepMerge(base[k], v)
        else
            base[k] = v
        end
    end
end

-- Insert nested value at a given path array into a table using deepMerge
local function insertAtPath(root, pathParts, value)
    if #pathParts == 0 then
        deepMerge(root, value)
        return
    end

    local current = root
    for i = 1, #pathParts - 1 do
        local part = pathParts[i]
        current[part] = current[part] or {}
        current = current[part]
    end

    local lastKey = pathParts[#pathParts]
    current[lastKey] = current[lastKey] or {}
    deepMerge(current[lastKey], value)
end

-- Serialize Lua table to string
local function serializeLuaTable(tbl, indent)
    indent = indent or ""
    local nextIndent = indent .. "  "
    local parts = {"{\n"}
    for k, v in pairs(tbl) do
        local key = string.format("[%q]", k)
        if type(v) == "table" then
            table.insert(parts, string.format("%s%s = %s,\n", nextIndent, key, serializeLuaTable(v, nextIndent)))
        else
            table.insert(parts, string.format("%s%s = %q,\n", nextIndent, key, v))
        end
    end
    table.insert(parts, indent .. "}")
    return table.concat(parts)
end

-- Recursively collect JSON files
local function collectFiles(path, rel, files)
    rel = rel or ""
    files = files or {}
    local fullPath = path .. (rel ~= "" and "/" .. rel or "")
    for _, entry in ipairs(listDir(fullPath)) do
        local subRel = rel ~= "" and (rel .. "/" .. entry) or entry
        local subFull = fullPath .. "/" .. entry
        if isDir(subFull) then
            collectFiles(path, subRel, files)
        elseif entry:match("^(%w+)%.json$") then
            table.insert(files, {
                lang = entry:match("^(%w+)"),
                path = path .. "/" .. subRel,
                relPath = subRel:match("(.+)/%w+%.json$") or ""
            })
        end
    end
    return files
end

-- Process JSON files
-- For each file, we assume the JSON structure is already nested and each leaf is an object with
-- keys "english" and "translation". For English, we extract the english text; for other languages, the translation.
local function buildLanguageTables()
    local allFiles = collectFiles(jsonRoot)
    local translations = {} -- lang -> table

    for _, file in ipairs(allFiles) do
        local lang = file.lang
        local relPathParts = {}
        for part in string.gmatch(file.relPath, "[^/]+") do
            table.insert(relPathParts, part)
        end

        local f = io.open(file.path, "r")
        local content = f:read("*a")
        f:close()

        local parsed = json.decode(content)
        -- Instead of flattening, use the nested structure directly.
        -- For each leaf, extract the appropriate field.
        local function processNode(node)
            if type(node) ~= "table" then
                return node
            end
            local result = {}
            for k, v in pairs(node) do
                if type(v) == "table" and (v.english or v.translation) then
                    -- Leaf node; choose the field based on language.
                    if lang == "en" then
                        result[k] = v.english or ""
                    else
                        result[k] = v.translation or ""
                    end
                else
                    result[k] = processNode(v)
                end
            end
            return result
        end

        local nested = processNode(parsed)
        translations[lang] = translations[lang] or {}
        insertAtPath(translations[lang], relPathParts, nested)
    end

    return translations
end

-- Main: write the final Lua files
local function writeAll()
    local translations = buildLanguageTables()
    ensureDir(outRoot)

    for lang, data in pairs(translations) do
        local outPath = outRoot .. "/" .. lang .. ".lua"
        local f = io.open(outPath, "w")
        f:write(fileHeader)
        f:write("return ")
        f:write(serializeLuaTable(data))
        f:close()
        print("Wrote:", outPath)
    end
end

writeAll()

File: bin/i18n/json-to-lua.py
import os
import json
import re
from collections import OrderedDict

JSON_ROOT = "json"
OUT_ROOT = "../../scripts/rfsuite/i18n"
HEADER_PATH = "lib/header.txt"


def read_file_header(path):
    if os.path.exists(path):
        with open(path, "r", encoding="utf-8") as f:
            return f.read() + "\n\n"
    return ""


def ensure_dir(path):
    os.makedirs(path, exist_ok=True)


def read_json_file(path):
    with open(path, "r", encoding="utf-8") as f:
        content = f.read()
        content = re.sub(r",\s*([}\]])", r"\1", content)  # Remove trailing commas
        return json.loads(content, object_pairs_hook=OrderedDict)


def lua_escape(value):
    value = str(value)
    value = value.replace('\\', '\\\\')  # escape backslashes
    value = value.replace('"', '\\"')     # escape double quotes
    value = value.replace('\n', '\\n')     # escape newlines
    value = value.replace('\r', '\\r')     # escape carriage returns
    return value


def serialize_lua_table(obj, indent="  ", level=0):
    if not isinstance(obj, dict):
        return f'"{lua_escape(obj)}"'

    lines = ["{"]
    keys = list(obj.keys())
    for i, key in enumerate(keys):
        val = obj[key]
        lua_key = f'[{json.dumps(key)}]'

        if isinstance(val, dict):
            lua_val = serialize_lua_table(val, indent, level + 1)
        elif isinstance(val, bool):
            lua_val = "true" if val else "false"
        elif isinstance(val, (int, float)):
            lua_val = str(val)
        else:
            lua_val = f'"{lua_escape(val)}"'

        comma = "," if i < len(keys) - 1 else ""
        lines.append(f'{indent * (level + 1)}{lua_key} = {lua_val}{comma}')

    lines.append(indent * level + "}")
    return "\n".join(lines)


def collect_json_files():
    json_files = []
    for root, _, files in os.walk(JSON_ROOT):
        for file in files:
            if file.endswith(".json"):
                lang = file[:-5]
                full_path = os.path.join(root, file)
                rel_path = os.path.relpath(full_path, JSON_ROOT)
                rel_dir = os.path.dirname(rel_path)
                json_files.append({
                    "lang": lang,
                    "path": full_path,
                    "rel_dir": rel_dir
                })
    return json_files


def insert_nested(target, rel_dir, value):
    parts = rel_dir.split(os.sep) if rel_dir else []
    current = target
    for part in parts:
        current = current.setdefault(part, OrderedDict())
    for k, v in value.items():
        current[k] = v


def process_json_content(lang, content):
    def extract(node):
        if isinstance(node, dict):
            result = OrderedDict()
            for k, v in node.items():
                if isinstance(v, dict) and ("english" in v or "translation" in v):
                    if lang == "en":
                        result[k] = v.get("english", "")
                    else:
                        result[k] = v.get("translation", "")
                else:
                    result[k] = extract(v)
            return result
        return node

    return extract(content)


def build_language_tables():
    translations = {}
    all_files = collect_json_files()

    for file_info in all_files:
        lang = file_info["lang"]
        rel_dir = file_info["rel_dir"]
        data = read_json_file(file_info["path"])
        processed = process_json_content(lang, data)
        translations.setdefault(lang, OrderedDict())
        insert_nested(translations[lang], rel_dir, processed)

    return translations


def write_lua_files():
    header = read_file_header(HEADER_PATH)
    translations = build_language_tables()
    ensure_dir(OUT_ROOT)

    for lang, data in translations.items():
        out_path = os.path.join(OUT_ROOT, f"{lang}.lua")
        with open(out_path, "w", encoding="utf-8") as f:
            f.write(header)
            f.write("return ")
            f.write(serialize_lua_table(data))
            f.write("\n")
        print("✔ Wrote:", out_path)


if __name__ == "__main__":
    write_lua_files()